### DISCUSSION TITLE
TypeError: Not() takes from 1 to 2 positional arguments but 3 were given

### DISCUSSION THREAD

**Discussion by 14H034160212** [Z3Prover/z3#5650] [Q&A]
Hi,

I got an issue when I try to use three elements inside the Not. Does anyone know how to solve it? Thank you so much.
![image](https://user-images.githubusercontent.com/23516191/140301306-df2def54-8830-461f-8fd4-b1eb08153507.png)

**Comment by aytey** [comment]
`Not` actually only takes _one_ argument (for what you want), the second argument is the context object, so doing `Not(p, q)` is going to give you closer to `Not(p)` than anything to do with `q`.

What do you hope the "single" result will be of `Not`ting multiple expressions would be anyway? A list?

Maybe you're trying to say that they're "not the same"? If so, you want `Distinct`.

**Comment by aytey** [comment]
> What do you hope the "single" result will be of `Not`ting multiple expressions would be anyway? A list?
> 
> Maybe you're trying to say that they're "not the same"? If so, you want `Distinct`.
>

Sorry, I didn't look at your screenshot properly.

If you do:

```python
s.solve(a, b, c)
```

that's the same as doing:

```python
s.solve(And([a, b, c]))
```

that is, it builds-up an implicit conjunction to solve all of the expressions you pass to `solve`.

I think what you want is going to be:

```python
s.solve(Not(And([a, b, c]))
# or
s.solve(Not(Or([a, b, c]))
```

if you want `And` or `Or` depends on if you want one expression to be falsified (so use `Not(And(...`, which becomes `Or(Not(...`) or *all* expressions to be falsified (so use `Not(Or(...)`, which becomes `And(Not(...`).

**Comment by 14H034160212** [comment]
> > What do you hope the "single" result will be of `Not`ting multiple expressions would be anyway? A list?
> > Maybe you're trying to say that they're "not the same"? If so, you want `Distinct`.
> 
> Sorry, I didn't look at your screenshot properly.
> 
> If you do:
> 
> ```python
> s.solve(a, b, c)
> ```
> 
> that's the same as doing:
> 
> ```python
> s.solve(And([a, b, c]))
> ```
> 
> that is, it builds-up an implicit conjunction to solve all of the expressions you pass to `solve`.
> 
> I think what you want is going to be:
> 
> ```python
> s.solve(Not(And([a, b, c]))
> # or
> s.solve(Not(Or([a, b, c]))
> ```
> 
> if you want `And` or `Or` depends on if you want one expression to be falsified (so use `Not(And(...`, which becomes `Or(Not(...`) or _all_ expressions to be falsified (so use `Not(Or(...)`, which becomes `And(Not(...`).

Thanks a lot for your quick reply. That make sense!
